Add basic function plotter plugin

rolandlo opened the pull request:
This is a basic function plotter using the recent Lua API extension by @WillNilges for adding strokes (and splines) to the Xournal++ document. For evaluating mathematical functions it uses the `luaxp`-module by Patrick Rigney. That module is included here (it's just one file and stands under the MIT license). 

The plugin supports parametric functions (x(t), y(t)). Taking x(t)=t will create ordinary functions y(t).

This is WIP. If someone would like to help or take this over, let me know.

Here is a quick demonstration:
![plotterDemo](https://user-images.githubusercontent.com/40485433/153748657-884b5e70-b556-4e35-ab11-772333801c9f.gif)


**Update**: Switched from the lgi module to LuaGObject (by default). On Windows and MacOS it comes bundled in the nightly release. On Linux you can install it via Luarocks:
``` term
luarocks install --lua-version=5.x LuaGObject
```

WillNilges commented:
Wow, very cool! Can this do frequency domain functions as well?

rolandlo commented:
> Wow, very cool! Can this do frequency domain functions as well?

Not yet. It's just 148 lines of Lua code, you can't expect too much of it. Glad you like it though. Looks like your implementation of `addStrokes` is pretty good. It satisfies all my needs so far.

Technius commented:
This is awesome and really shows off the power of plugins. Only 147 lines of Lua for the main logic!

Photon89 commented:
That's fantastic, when will it get released? :smiley:

rolandlo commented:
@Photon89 It will probably take a while, since I am currently working on the .note file import plugin. If you have a recent nightly build installed, you can already use it by copying all the files into a subfolder of the plugin folder (`usr/share/xournalpp/plugins` on Linux) though.

Photon89 commented:
I see, thanks! I guess, I will create an AUR package in a few weeks, when I have some spare time!

tattsan commented:
Nice updates!  It used to be inconvenient to use the pow function to represent powers.

rolandlo commented:
> Nice updates! It used to be inconvenient to use the pow function to represent powers, but now we can use ^ operator. Thank you so much!

You're welcome! :smile:  Nice to get some feedback.

I have pushed more updates. Now Euler's number can be used as well (e.g. "e^t"). Moreover there is an appearance tab that lets the user choose what to draw (axis lines, ticks, grid) and in what color, line width and line style.

"Display numbers" (near ticks) and "Auto-fit" (for coordinate window) are not implemented yet. The former will require an extension of the plugin API.

mjg commented:
Just tried this out for the 1st time (missed this teaching today), and it is grrreat! Thanks a bunch. Will add this branch to my own builds ;-)

Due to the implementation as stroke generator, one can use the WIP plugin without creating potentially incompatible xopps. Perfect!

myNicole98 commented:
Hey there, I was trying this plugin but when I try to plot something It raises this error:

`Error raised while calling 'lgi.cbk (function: 0x55b9ac762bd0): GObject': /usr/share/xournalpp/plugins/FunctionPlotter/main.lua:194: attempt to call a nil value (field 'addStrokes')`

Any idea on how can I fix this?

myNicole98 commented:
I'm using version 1.1.3

myNicole98 commented:
Yup, now it's working, I installed the git version, nice plugin!

isaacalbert commented:
Thanks for this great plugin.
I used it in my math class last year.
How can i use the "display numbers" option? It is possible to select, but no numbers on the axis.
I looked in the code and it seems that it is not implemented, or?
What do you think, what effort does it need with basic code experience?
Thanks.

Photon89 commented:
So I finally found some time to test it and create an AUR package: https://aur.archlinux.org/packages/xournalpp-plugin-functionplotter-git

I wonder, if it is possible to add axis labels next to the axis arrows (like "x" and "y") to the appearance options?

rolandlo commented:
> Thanks for this great plugin. I used it in my math class last year. How can i use the "display numbers" option? It is possible to select, but no numbers on the axis. I looked in the code and it seems that it is not implemented, or? What do you think, what effort does it need with basic code experience? Thanks.

You're correct, "display numbers" is not implemented and the plugin is unfinished. The plugin API has to be extended by `app.addText` (and `app.getText`), in a PR separate from this plugin. It requires some familiarity with the interaction between Lua and C++, but it should be relatively straightforward. Compare #4884.
When `app.addText` is available it would also be easy to add axis labels next to the axis arrows and make that part of the appearance options.

rolandlo commented:
Display numbers now works. Here e.g. we have x(t)=t,  y(t)=cos(pi*t)

![grafik](https://github.com/xournalpp/xournalpp/assets/40485433/bb7cc77b-e287-4ba5-b598-102880ef879d)

You will need the nightly build though, since the API for adding text was just merged.

isaacalbert commented:
Great job!
That's what i need at the new school year an for the advanced training for the teachers next week!
To implement the add.text was too difficult for me. 
I have chosen the brute force method and added numbers with strokes from a xournal++ file.
The advantage is that i can edit the coordinate system afterwards with the eraser and in the math lessons we need centimeters because of the graph templates we use.
So i have to implement another button to use both number styles...
Thanks a lot!

https://github.com/xournalpp/xournalpp/assets/87916000/5afb942f-32bb-40ff-a338-310a5b3e0dee

atticus-sullivan commented:
> That's what i need at the new school year an for the advanced training for the teachers next week!

You're training teachers to work with tools on Linux? That's pretty cool (well if they are mostly using xournalpp as a whiteboard, the OS on the outside does not matter)

> So i have to implement another button to use both number styles...

Of what different number styles are you talking?

rolandlo commented:
> I have chosen the brute force method and added numbers with strokes from a xournal++ file.

That sounds like an interesting approach, that could be used to provide a shape library. Can you show that code how you did that?

isaacalbert commented:
> You're training teachers to work with tools on Linux? That's pretty cool (well if they are mostly using xournalpp as a whiteboard, the OS on the outside does not matter)

Yes but its a hard restrictiv microsoft regime in my school. Perhaps i can convince some teachers to use xournal on the tablets and smartboards.

> Of what different number styles are you talking?

The one with add.text that rolandlo implemented and my add.stoke version so that all elements in the coordinate system are strokes.


> That sounds like an interesting approach, that could be used to provide a shape library. Can you show that code how you did that?

I just draw the numbers on the magnetic dotted background in xournal, extract the coordinates, arrange them with LO Calc and put them into main.lua.
I changed some default values in plotter.glade for the coordinate system with centimeters, so that it is exact like the students draw.

[123xy.xopp.txt](https://github.com/xournalpp/xournalpp/files/12381164/123xy.xopp.txt)
[main.lua.txt](https://github.com/xournalpp/xournalpp/files/12381208/main.lua.txt)
[plotter.glade.txt](https://github.com/xournalpp/xournalpp/files/12381219/plotter.glade.txt)

with best regards

rolandlo commented:
With the last commits I have added optional axis labels and a font chooser button to choose the font for the numbers and axis labels.
Deprecated `app.msgbox` was replaced by `app.openDialog`. Also rebased on master.

mjg commented:
> Sorry to ask here, where can I download the code of the plotter for Xournalpp? Like it to use for school but haven't found the code yet. I don't use Arch btw ;) Using Fedora doesn't matter?
> 
> Regards

I'd say both Fedora and Arch are great choices ;-)

You could just download the plugin parts, but I seem to recall that they need a recent xournalpp. If you are in desperate need to get the plotter you are welcome to use my Fedora builds (`dnf copr enable mjg/xournalpp`) for x86_64. But please be aware that they are my personal test builds (based on upstream/master) and may contain additional patches (currently rolandlo's plotter branch and my copy-paste-selection-anchors). No warranties but - during the term - I use that build regularly myself. I'm not the maintainer of the official Fedora package for xournalpp (but of other ones).

rolandlo commented:
> Sorry to ask here, where can I download the code of the plotter for Xournalpp? Like it to use for school but haven't found the code yet. I don't use Arch btw ;) Using Fedora doesn't matter?
> 
> Regards

You find the code under the [Files changed tab](https://github.com/xournalpp/xournalpp/pull/3811/files).
On Arch there is already a package by @Photon89, see [this post](https://github.com/xournalpp/xournalpp/pull/3811#issuecomment-1670944171).

The current code makes use of `app.openDialog` introduced in #5079 on Aug 23. Also `app.addTexts` from #5029 is used (merged on Aug 17).  So you really need a recent build of Xournal++ version `1.2.1+dev` if you want to run the code unmodified. 

Let me remind everyone that this is unfinished work (although close to finished now, since all basic features for a function plotter are implemented) and it's very possible that some temporary regressions are introduced while working on it.

nonameboy-opi commented:
Thanks for the answers the plotter is working very good in xounalpp :)

Photon89 commented:
Works very well for me as well! I wonder, whether it is possible to group all objects that constitute the graph to one graph object, that can be handled (moved, resized, edited) as a single object. Most probably that's beyond what is possible with the current plugin interface though...

kaliasgoldmedal commented:
I have copied the plugin and pasted in the installed folder. When I try to use the plugin I get an error as below

You should have Lua lgi-module installed and include in your lua package path.

Help me.

atticus-sullivan commented:
> Works very well for me as well! I wonder, whether it is possible to group all objects that constitute the graph to one graph object, that can be handled (moved, resized, edited) as a single object. Most probably that's beyond what is possible with the current plugin interface though...

I think this is more of an issue xournalpp doesn't allow grouping multiple objects together (afaik). A workaround might be putting the plot to a new layer. This is currently possible.

rolandlo commented:
> I have copied the plugin and pasted in the installed folder. When I try to use the plugin I get an error as below
> 
> You should have Lua lgi-module installed and include in your lua package path.
> 
> Help me.

As I remember getting the `lgi` module to run on Windows is not as straight forward as on Linux, You might try to install `luarocks` as in https://github.com/luarocks/luarocks/wiki/Installation-instructions-for-Windows and then install `lgi` through luarocks https://luarocks.org/modules/pavouk/lgi

rolandlo commented:
@kaliasgoldmedal 
I managed to install `lgi` on Windows. See https://github.com/xournalpp/xournalpp/discussions/4522 for how to do it.
The plugin works now on Windows, after I fixed the path separator with the latest commit.

Prola3 commented:
@rolandlo caveman question, how to add this plugin to the program, Linux mint user

rolandlo commented:
> @rolandlo caveman question, how to add this plugin to the program, Linux mint user

First of all you need the nightly build (1.2.5+dev), not the stable version (1.2.5). Then in case you haven't set up git the easiest way is to just download the zip-file retrieved by clicking on this link: https://api.github.com/repos/xournalpp/xournalpp/zipball/pull/3811/head and extract the subfolder `plugins/FunctionPlotter` into the folder `~/.config/xournalpp/plugins`.
